Heap buffer overflow in opj_t1_clbl_decode_processor() triggered with Ghostscript

Version: `2.3.0`

Command to reproduce (Ghostscript): `gs -dNOPAUSE -sDEVICE=bit -sOUTPUTFILE=/dev/null -dSAFER gs_hbo_opj_t1_clbl_decode_processor -c quit`

Crashing test case (please unpack): [gs_hbo_opj_t1_clbl_decode_processor.zip](https://github.com/uclouvain/openjpeg/files/2495501/gs_hbo_opj_t1_clbl_decode_processor.zip)


ASAN log:

```
==5834==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x7f56033c0440 at pc 0x55beea74fe6c bp 0x7ffd95c8bfa0 sp 0x7ffd95c8bf90
WRITE of size 4 at 0x7f56033c0440 thread T0
    #0 0x55beea74fe6b in opj_t1_clbl_decode_processor openjpeg/src/lib/openjp2/t1.c:1782
    #1 0x55beea7a6205 in opj_thread_pool_submit_job openjpeg/src/lib/openjp2/thread.c:833
    #2 0x55beea753023 in opj_t1_decode_cblks openjpeg/src/lib/openjp2/t1.c:1901
    #3 0x55beea7912e6 in opj_tcd_t1_decode openjpeg/src/lib/openjp2/tcd.c:1963
    #4 0x55beea7912e6 in opj_tcd_decode_tile openjpeg/src/lib/openjp2/tcd.c:1617
    #5 0x55beea68c150 in opj_j2k_decode_tile openjpeg/src/lib/openjp2/j2k.c:8935
    #6 0x55beea68c150 in opj_j2k_decode_tiles openjpeg/src/lib/openjp2/j2k.c:10722
    #7 0x55beea643c8f in opj_j2k_exec openjpeg/src/lib/openjp2/j2k.c:8096
    #8 0x55beea699500 in opj_j2k_decode openjpeg/src/lib/openjp2/j2k.c:11017
    #9 0x55beea6b4a44 in opj_jp2_decode openjpeg/src/lib/openjp2/jp2.c:1604
    #10 0x55beea5e9a45 in decode_image base/sjpx_openjpeg.c:407
    #11 0x55beea5e9a45 in s_opjd_process base/sjpx_openjpeg.c:734
    #12 0x55beea9d3b79 in sreadbuf base/stream.c:823
    #13 0x55beea9e4a98 in s_process_read_buf base/stream.c:749
    #14 0x55beec45367a in image_file_continue psi/zimage.c:533
    #15 0x55beec2751b8 in interp psi/interp.c:1256
    #16 0x55beec2751b8 in gs_call_interp psi/interp.c:516
    #17 0x55beec284c0d in gs_interpret psi/interp.c:473
    #18 0x55beec221e2f in gs_main_interpret psi/imain.c:235
    #19 0x55beec221e2f in gs_main_run_string_end psi/imain.c:658
    #20 0x55beec221e2f in gs_main_run_string_with_length psi/imain.c:610
    #21 0x55beec221e2f in gs_main_run_string psi/imain.c:591
    #22 0x55beec22f0e8 in run_string psi/imainarg.c:1034
    #23 0x55beec22f0e8 in runarg psi/imainarg.c:1024
    #24 0x55beec23898b in argproc psi/imainarg.c:957
    #25 0x55beec23898b in gs_main_init_with_args psi/imainarg.c:233
    #26 0x55bee9a9e249 in main psi/gs.c:95
    #27 0x7f561eaebb96 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b96)
    #28 0x55bee9ab3289 in _start (XYZ/gs_asan/bin/gs+0x36a289)

0x7f56033c0441 is located 0 bytes to the right of 261185-byte region [0x7f5603380800,0x7f56033c0441)
allocated by thread T0 here:
    #0 0x7f56203d5b50 in __interceptor_malloc (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xdeb50)
    #1 0x55beeb7bb9c9 in gs_heap_alloc_bytes base/gsmalloc.c:193

SUMMARY: AddressSanitizer: heap-buffer-overflow openjpeg/src/lib/openjp2/t1.c:1782 in opj_t1_clbl_decode_processor
Shadow bytes around the buggy address:
  0x0feb40670030: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0feb40670040: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0feb40670050: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0feb40670060: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0feb40670070: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x0feb40670080: 00 00 00 00 00 00 00 00[01]fa fa fa fa fa fa fa
  0x0feb40670090: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0feb406700a0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0feb406700b0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0feb406700c0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0feb406700d0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==5834==ABORTING
```

I've extracted the JP2 content from the .gs file (
[gs_hbo_opj_t1_clbl_decode_processor_jp2.jp2.zip](https://github.com/uclouvain/openjpeg/files/2495664/gs_hbo_opj_t1_clbl_decode_processor_jp2.jp2.zip) )
, and used it with opj_decompress with a -fsanitize=address build of openjpeg v2.3.0 and master, and I don't get any corruption
```
$ bin/opj_decompress -i  gs_hbo_opj_t1_clbl_decode_processor_jp2.jp2 -o out.tif

[INFO] JP2 IHDR box: compression type indicate that the file is not a conforming JP2 file (48) 
[INFO] Start to read j2k main header (85).
[WARNING] Despite JP2 BPC!=255, precision and/or sgnd values for comp[1] is different than comp[0]:
        [0] prec(8) sgnd(0) [1] prec(8) sgnd(1)
[INFO] Main header has been correctly decoded.
[INFO] No decoded area parameters, set the decoded area to the whole image
[INFO] Psot value of the current tile-part is equal to zero, we assuming it is the last tile-part of the codestream.
[ERROR] Tile part length size inconsistent with stream length
[ERROR] Failed to decode the codestream in the JP2 file
ERROR -> opj_decompress: failed to decode image!
```

--- a/src/bin/jp2/opj_decompress.c
+++ b/src/bin/jp2/opj_decompress.c
@@ -1571,17 +1571,6 @@ int main(int argc, char **argv)
             }
         }
 
-        /* FIXME? Shouldn't that situation be considered as an error of */
-        /* opj_decode() / opj_get_decoded_tile() ? */
-        if (image->comps[0].data == NULL) {
-            fprintf(stderr, "ERROR -> opj_decompress: no image data!\n");
-            opj_destroy_codec(l_codec);
-            opj_stream_destroy(l_stream);
-            opj_image_destroy(image);
-            failed = 1;
-            goto fin;
-        }
-
         tCumulative += opj_clock() - t;
         numDecompressedImages++;
 

--- a/src/lib/openjp2/j2k.c
+++ b/src/lib/openjp2/j2k.c
@@ -10657,6 +10657,42 @@ static OPJ_BOOL opj_j2k_allocate_tile_element_cstr_index(opj_j2k_t *p_j2k)
     return OPJ_TRUE;
 }
 
+static OPJ_BOOL opj_j2k_are_all_used_components_decoded(opj_j2k_t *p_j2k,
+        opj_event_mgr_t * p_manager)
+{
+    OPJ_UINT32 compno;
+    OPJ_BOOL decoded_all_used_components = OPJ_TRUE;
+
+    if (p_j2k->m_specific_param.m_decoder.m_numcomps_to_decode) {
+        for (compno = 0;
+                compno < p_j2k->m_specific_param.m_decoder.m_numcomps_to_decode; compno++) {
+            OPJ_UINT32 dec_compno =
+                p_j2k->m_specific_param.m_decoder.m_comps_indices_to_decode[compno];
+            if (p_j2k->m_output_image->comps[dec_compno].data == NULL) {
+                opj_event_msg(p_manager, EVT_WARNING, "Failed to decode component %d\n",
+                              dec_compno);
+                decoded_all_used_components = OPJ_FALSE;
+            }
+        }
+    } else {
+        for (compno = 0; compno < p_j2k->m_output_image->numcomps; compno++) {
+            if (p_j2k->m_output_image->comps[compno].data == NULL) {
+                opj_event_msg(p_manager, EVT_WARNING, "Failed to decode component %d\n",
+                              compno);
+                decoded_all_used_components = OPJ_FALSE;
+            }
+        }
+    }
+
+    if (decoded_all_used_components == OPJ_FALSE) {
+        opj_event_msg(p_manager, EVT_ERROR, "Failed to decode all used components\n");
+        return OPJ_FALSE;
+    }
+
+    return OPJ_TRUE;
+}
+
+
 static OPJ_BOOL opj_j2k_decode_tiles(opj_j2k_t *p_j2k,
                                      opj_stream_private_t *p_stream,
                                      opj_event_mgr_t * p_manager)
@@ -10766,6 +10802,10 @@ static OPJ_BOOL opj_j2k_decode_tiles(opj_j2k_t *p_j2k,
         if (++nr_tiles ==  p_j2k->m_cp.th * p_j2k->m_cp.tw) {
             break;
         }
+    }
+
+    if (! opj_j2k_are_all_used_components_decoded(p_j2k, p_manager)) {
+        return OPJ_FALSE;
     }
 
     return OPJ_TRUE;
@@ -10894,6 +10934,10 @@ static OPJ_BOOL opj_j2k_decode_one_tile(opj_j2k_t *p_j2k,
                           l_current_tile_no + 1, l_tile_no_to_dec + 1);
         }
 
+    }
+
+    if (! opj_j2k_are_all_used_components_decoded(p_j2k, p_manager)) {
+        return OPJ_FALSE;
     }
 
     return OPJ_TRUE;

